airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/manifest_component_transformer.py

```diff
@@ -0,0 +1,110 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+import copy
+import typing
+from typing import Any, Mapping
+
+OPTIONS_STR = "$options"
+
+
+# todo: For better granularity, we may want this to be keyed on the object + field
+DEFAULT_MODEL_TYPES: Mapping[str, str] = {
+    # DatetimeStreamSlicer
+    "end_datetime": "MinMaxDatetime",
+    "start_datetime": "MinMaxDatetime",
+    # DeclarativeSource
+    "streams": "DeclarativeStream",
+    # DeclarativeStream
+    "retriever": "SimpleRetriever",
+    "schema_loader": "DefaultSchemaLoader",
+    # CursorPagination, DefaultPaginator, DpathExtractor
+    "decoder": "JsonDecoder",
+    # DefaultErrorHandler
+    "response_filters": "HttpResponseFilter",
+    # DefaultPaginator
+    "page_size_option": "RequestOption",
+    "page_token_option": "RequestOption",
+    # HttpRequester
+    "error_handler": "DefaultErrorHandler",
+    "request_options_provider": "InterpolatedRequestOptionsProvider",
+    # ListStreamSlicer, ParentStreamConfig
+    "request_option": "RequestOption",
+    # ParentStreamConfig
+    "stream": "DeclarativeStream",
+    # RecordSelector
+    "extractor": "DpathExtractor",
+    # SimpleRetriever
+    "paginator": "NoPagination",
+    "record_selector": "RecordSelector",
+    "requester": "HttpRequester",
+    "stream_slicer": "SingleSlice",
+    # SubstreamSlicer
+    "parent_stream_configs": "ParentStreamConfig",
+    # Transformers
+    "fields": "AddedFieldDefinition",
+}
+
+# We retain a separate registry for custom components to automatically insert the type if it is missing. This is intended to
+# be a short term fix because once we have migrated, then type and class_name should be requirements for all custom components.
+CUSTOM_COMPONENTS_MAPPING: Mapping[str, str] = {
+    "authenticator": "CustomAuthenticator",
+    "backoff_strategies": "CustomBackoffStrategy",
+    "extractor": "CustomRecordExtractor",
+    "pagination_strategy": "CustomPaginationStrategy",
+    "stream_slicer": "CustomStreamSlicer",
+}
+
+
+class ManifestComponentTransformer:
```

Comment:
As mentioned in the description, we need this transformers in the immediate term because we do a serious of tweaks to the manifest prior to validating the schema like passing options to all subcomponents and appending types based on the old `DEFAULT_IMPLEMENTATION_REGISTY`. I do think that in the longer term we would want to always enforce a `type` be defined on all components, but in the moment, so many connectors don't have types set that we need this intermediate "hydrator" to exist since we no longer want to rely on the factory